repository/src/lib.rs

```diff
@@ -415,13 +408,103 @@ impl<T: RawRepository> DistributedRepository<T> {
+        if new_semantic_commits.is_err() {
+            log::warn!("Found invalid remote tracking branches");
+        }
+        let new_semantic_commits = new_semantic_commits.unwrap();
+        let new_commits = new_semantic_commits
+            .into_iter()
+            .map(|(commit, hash)| {
+                from_semantic_commit(commit)
+                    .map_err(|e| (e, hash))
+                    .map(|x| (x, hash))
+            })
+            .collect::<Result<Vec<_>, _>>();
+        if new_commits.is_err() {
+            log::warn!("Found invalid remote tracking branches");
+        }
+        let new_commits = new_commits.unwrap();
+
+        let start_header = self.get_last_finalized_block_header().await?;
+        let reserved_state = self.get_reserved_state().await?;
+
+        // Verify every commits along the way.
+        let mut verifier =
+            CommitSequenceVerifier::new(start_header.clone(), reserved_state.clone())
+                .map_err(|e| anyhow!("failed to create a commit sequence verifier: {}", e))?;
+        for (new_commit, new_commit_hash) in &new_commits {
+            verifier
+                .apply_commit(new_commit)
+                .map_err(|e| anyhow!("verification error on commit {}: {}", new_commit_hash, e))?;
+        }
+
+        // Verify finalization proof
+        let fp_commit_hash = self.raw.locate_branch(FP_BRANCH_NAME.into()).await?;
+        let fp_semantic_commit = self.raw.read_semantic_commit(fp_commit_hash).await?;
+        let finalization_proof: FinalizationProof =
+            serde_json::from_str(&fp_semantic_commit.body).unwrap();
+        let block_semantic_commit = self.raw.read_semantic_commit(*block_commit_hash).await?;
+        let block_commit =
+            format::from_semantic_commit(block_semantic_commit).map_err(|e| anyhow!(e))?;
+        if let Commit::Block(block_header) = block_commit {
+            verify::verify_finalization_proof(&block_header, &finalization_proof)
+                .map_err(|e| anyhow!("failed to verify finalization proof: {}", e))?;
+        }
+
+        // If commit sequence verification is done and the finalization proof is verified,
+        // move the `finalized` branch to the given block commit hash.
+        // Then we update the `fp` branch.
+        self.raw.checkout_clean().await?;
+        let last_finalization_proof = format::fp_from_semantic_commit(
+            self.raw
+                .read_semantic_commit(self.raw.locate_branch(FP_BRANCH_NAME.into()).await?)
+                .await?,
+        )?;
+        self.raw
+            .move_branch(FINALIZED_BRANCH_NAME.to_string(), *block_commit_hash)
+            .await?;
+        self.raw
+            .move_branch(FP_BRANCH_NAME.into(), *block_commit_hash)
+            .await?;
+        self.raw.checkout(FP_BRANCH_NAME.into()).await?;
+        self.raw
+            .create_semantic_commit(format::fp_to_semantic_commit(&last_finalization_proof))
+            .await?;
```

Comment:
in this way, aren't there multiple fp commits on the `fp` branch?

Reply:
fixed.

Reply:
how did you fix it? I can't see the difference

Reply:
I thought that finalization_proof is only one. Before changing it it was last_finalization_proof so we have Vector of proofs. Can you check it?